source/common/conn_pool/conn_pool_base.cc

```diff
@@ -449,22 +493,26 @@ void ConnPoolImplBase::onConnectionEvent(ActiveClient& client, absl::string_view
       host_->cluster().stats().upstream_cx_connect_fail_.inc();
       host_->stats().cx_connect_fail_.inc();
 
-      ConnectionPool::PoolFailureReason reason;
-      if (client.timed_out_) {
-        reason = ConnectionPool::PoolFailureReason::Timeout;
-      } else if (event == Network::ConnectionEvent::RemoteClose) {
-        reason = ConnectionPool::PoolFailureReason::RemoteConnectionFailure;
-      } else {
-        reason = ConnectionPool::PoolFailureReason::LocalConnectionFailure;
+      onConnectFailed(client);
+      if (contributed_to_connecting_stream_capacity) {
+        // Purge pending streams only if this is not a.
+        ConnectionPool::PoolFailureReason reason;
+        if (client.timed_out_) {
+          reason = ConnectionPool::PoolFailureReason::Timeout;
+        } else if (event == Network::ConnectionEvent::RemoteClose) {
+          reason = ConnectionPool::PoolFailureReason::RemoteConnectionFailure;
+        } else {
+          reason = ConnectionPool::PoolFailureReason::LocalConnectionFailure;
+        }
+
+        // Raw connect failures should never happen under normal circumstances. If we have an
+        // upstream that is behaving badly, streams can get stuck here in the pending state. If we
+        // see a connect failure, we purge all pending streams so that calling code can determine
+        // what to do with the stream.
+        // NOTE: We move the existing pending streams to a temporary list. This is done so that
+        //       if retry logic submits a new stream to the pool, we don't fail it inline.
+        purgePendingStreams(client.real_host_description_, failure_reason, reason);
```

Comment:
As per my comment on line 477, I don't think I understand why we're not doing this when the unused capacity is 0. But I'm sure you do :)